File: 14_binary_search_algorithm/min_max_division.py
def check(val, k, a):
  this = 0;
  for i in a:
    if i > val:
      return 0
    this += i
    if (this > val):
      k -= 1
      this = i
  return k > 0

def solution(k, m, a):
  n = len(a)
  beg = 0
  end = m * n
  rt = -1
  while beg <= end:
    mid = (beg + end) / 2
    if check(mid, k, a):
      end = mid - 1
      rt = mid;
    else:
      beg = mid + 1
  return rt

File: 14_binary_search_algorithm/min_max_division.cpp
#include <vector>

using namespace std;

bool check(int val, int k, vector<int> &a) {
  int this_sum = 0;
  int n = a.size();
  for (int i = 0; i < n; i++) {
    if (a[i] > val) {
      return 0;
    }
    this_sum += a[i];
    if (this_sum > val) {
      k--;
      this_sum = a[i];
    }
  }
  return (k > 0);
}

int solution(int k, int m, vector<int> &a) {
  int n = a.size();
  int beg = 0;
  int end = m * n;
  int mid;
  int rt = -1;
  while (beg <= end) {
    mid = (beg + end) / 2;
    if (check(mid, k, a)) {
      end = mid - 1;
      rt = mid;
    }
    else {
      beg = mid + 1;
    }
  }
  return rt;
}
